### PR TITLE
fix: ensure headers can be set and overriden

### PR DESCRIPTION
**Pull request by ewanharris**
#### What problem is being solved?

The project does not document or test custom headers support

#### How is it being solved?

Documentation and tests added, fixes made where relevant

#### What changes are made to solve it?

Adds tests and documentation for setting default headers and overriding them per-request, this also extended to a few fixes due to issue uncovered, they are:

* c4b416fe2064b0f71b5162622c9448d03bd1dbcd - Overriding `defaultHeaders` on `ClientConfiguration` so typing works correctly
* 705472f56239ff8ab0ce8d2c7aa4342612c30c42 - Correctly handles calling `readAuthorizationModel` with options but no model ID set
* 67a625e7f3bcc8caddbc835ce71875ba70492aff - Sets the headers when copying over to the `ClientBatchCheck` options

There is one discussion point I have that I will comment in the PR regarding

## References

https://github.com/openfga/sdk-generator/issues/569

## Review Checklist
- [x] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [x] The correct base branch is being used, if not `main`
- [x] I have added tests to validate that the change in functionality is working as expected

## Summary by CodeRabbit

- New Features
  - Fluent configuration for default HTTP headers in the client.

- Bug Fixes
  - Prefer per-call authorization model when provided; otherwise fall back to configuration.
  - Copy per-request header maps to prevent unintended mutations.

- Documentation
  - Added README examples for default and per-request headers (note: content was inserted twice).

- Tests
  - Added comprehensive header tests covering merging, overrides, edge cases, and request verification.

### CODE REVIEW

File: src/main/java/dev/openfga/sdk/api/configuration/ClientListRelationsOptions.java [hunk 1]
```diff
@@ -61,6 +61,7 @@ public ConsistencyPreference getConsistency() {
     public ClientBatchCheckClientOptions asClientBatchCheckClientOptions() {
         return new ClientBatchCheckClientOptions()
                 .authorizationModelId(authorizationModelId)
+                .additionalHeaders(additionalHeaders)
                 .maxParallelRequests(maxParallelRequests)
                 .consistency(consistency);
```
[the remark this answers is not in the document]
Comment (ewanharris) [reply src/main/java/dev/openfga/sdk/api/configuration/ClientListRelationsOptions.java:67]: This was added

File: src/test/java/dev/openfga/sdk/api/client/OpenFgaClientHeadersTest.java [hunk 2]
```diff
@@ -0,0 +1,839 @@
...
+                .withHeader("another-header", "another-value")
+                .withHeader("test-header", "test-value-per-call")
+                .doReturn(200, "{\"result\":{}}");
+        ClientBatchCheckItem item = new ClientBatchCheckItem()
+                .user(DEFAULT_USER)
+                .relation(DEFAULT_RELATION)
+                ._object(DEFAULT_OBJECT)
+                .correlationId("cor-1");
+        ClientBatchCheckRequest request = new ClientBatchCheckRequest().checks(List.of(item));
+        // Use HashMap instead of Map.of() to create a mutable map
```
[the remark this answers is not in the document]
Comment (jimmyjames) [reply src/test/java/dev/openfga/sdk/api/client/OpenFgaClientHeadersTest.java:483]: Yes, I agree it is confusing to do it with `Map.of` everywhere else but here. Let's keep it consistent and use `Map.of` here too if we can.

[the remark this answers is not in the document]
Comment (ewanharris) [reply src/test/java/dev/openfga/sdk/api/client/OpenFgaClientHeadersTest.java:483]: Done in d478606d8c3a692f5799605cb3a5d22c65687b2d and updated the tests to use `Map.of` so that validates the behaviour

**Comment by ewanharris** [comment]
@coderabbitai review

File: src/main/java/dev/openfga/sdk/api/client/OpenFgaClient.java [hunk 3]
```diff
@@ -1130,12 +1131,13 @@ public CompletableFuture<ClientListRelationsResponse> listRelations(
...
-        }
-        options.getAdditionalHeaders().putIfAbsent(CLIENT_METHOD_HEADER, "ListRelations");
-        options.getAdditionalHeaders()
-                .putIfAbsent(CLIENT_BULK_REQUEST_ID_HEADER, randomUUID().toString());
+
+        HashMap<String, String> headers = options.getAdditionalHeaders() != null
+                ? new HashMap<>(options.getAdditionalHeaders())
+                : new HashMap<>();
+        headers.put(CLIENT_METHOD_HEADER, "ListRelations");
+        headers.put(CLIENT_BULK_REQUEST_ID_HEADER, randomUUID().toString());
```
Comment (jimmyjames) [inline src/main/java/dev/openfga/sdk/api/client/OpenFgaClient.java:1139]: Why `put` here but `putIfAbsent` above?

Comment (ewanharris) [reply src/main/java/dev/openfga/sdk/api/client/OpenFgaClient.java:1139]: As `listRelations` calls `clientBatchCheck` we don't want to stomp over the existing values, agree it's kinda awkward having the two differ so maybe I should just match the two?

**Comment by ewanharris** [comment]
@coderabbitai review

**Review by jimmyjames — APPROVED** [review]
